src/controller/AutoCommissioner.cpp

```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+            if (mParams.GetExecuteJCM().ValueOr(false) &&
+                (mDeviceCommissioningInfo.JFAdministratorFabricIndex != kUndefinedFabricIndex))
+            {
+                if (!mJFAdminRCAC.Alloc(mDeviceCommissioningInfo.JFAdminRCAC.size()))
+                {
+                    ChipLogError(Controller, "AutoCommissioner: cannot allocate memory for mJFAdminRCAC");
+                    return CHIP_ERROR_NO_MEMORY;
+                    ;
+                }
+                memcpy(mJFAdminRCAC.Get(), mDeviceCommissioningInfo.JFAdminRCAC.data(), mJFAdminRCAC.AllocatedSize());
+                mParams.SetJFAdminRCAC(ByteSpan(mJFAdminRCAC.Get(), mJFAdminRCAC.AllocatedSize()));
+
+                if (!mJFAdminICAC.Alloc(mDeviceCommissioningInfo.JFAdminICAC.size()))
+                {
+                    ChipLogError(Controller, "AutoCommissioner: cannot allocate memory for mJFAdminICAC");
+                    return CHIP_ERROR_NO_MEMORY;
+                    ;
+                }
+                memcpy(mJFAdminICAC.Get(), mDeviceCommissioningInfo.JFAdminICAC.data(), mJFAdminICAC.AllocatedSize());
+                mParams.SetJFAdminICAC(ByteSpan(mJFAdminICAC.Get(), mJFAdminICAC.AllocatedSize()));
+
+                if (!mJFAdminNOC.Alloc(mDeviceCommissioningInfo.JFAdminNOC.size()))
+                {
+                    ChipLogError(Controller, "AutoCommissioner: cannot allocate memory for mJFAdminNOC");
+                    return CHIP_ERROR_NO_MEMORY;
+                    ;
+                }
+                memcpy(mJFAdminNOC.Get(), mDeviceCommissioningInfo.JFAdminNOC.data(), mJFAdminNOC.AllocatedSize());
+                mParams.SetJFAdminNOC(ByteSpan(mJFAdminNOC.Get(), mJFAdminNOC.AllocatedSize()));
+
+                mParams.SetJFAdminEndpointId(mDeviceCommissioningInfo.JFAdminEndpointId)
+                    .SetJFAdministratorFabricIndex(mDeviceCommissioningInfo.JFAdministratorFabricIndex)
+                    .SetJFAdminFabricId(mDeviceCommissioningInfo.JFAdminFabricTable.fabricId)
+                    .SetJFAdminVendorId(mDeviceCommissioningInfo.JFAdminFabricTable.vendorId);
+            }
+#endif // CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
+
             break;
         }
         case CommissioningStage::kConfigureTimeZone:
             mNeedsDST = report.Get<TimeZoneResponseInfo>().requiresDSTOffsets;
             break;
-        case CommissioningStage::kSendPAICertificateRequest:
-            SetPAI(report.Get<RequestedCertificate>().certificate);
+        case CommissioningStage::kSendPAICertificateRequest: {
+            auto reportPAISpan = report.Get<RequestedCertificate>().certificate;
+
+            if (!mPAI.Alloc(reportPAISpan.size()))
+            {
+                ChipLogError(Controller, "AutoCommissioner cannot allocate memory for mPAI");
+                return CHIP_ERROR_NO_MEMORY;
+            }
+
+            memcpy(mPAI.Get(), reportPAISpan.data(), mPAI.AllocatedSize());
```

Comment:
We have a bunch of copies of this code.  Can we put it into a helper function, please?

Reply:
Yes, I created AllocateMemoryAndCopySpan() helper function.